### ISSUE TITLE
CSI Driver not registering

### ISSUE THREAD

**Issue by starcraft66** [democratic-csi/democratic-csi#348]
Hello, I've migrated my kubernetes cluster to a different operating system (used to be on ubuntu VMs and am now on NixOS bare metal) the source cluster and destination cluster are both v1.27.2 and I'm using the democratic-csi helm chart 0.14.2.

Ever since then I haven't been able to get democratic-csi to work. The controllers and node daemons are crash looping every minute or so. 

EDIT: I'm pretty sure the controller crash looping is unrelated, it's much less frequent and seems to be related to an intermittent connectivity issue caused by cilium crashing and dropping BGP sessions with my router resulting in the truenass ssh server being unreachable, but I don't think this driver registration problem is related because it manifests even when the controller has ssh access to truenas and I can see it logging in via truenas' /var/log/auth.log.

```
❯ kgpon democratic-csi
NAME                                                              READY   STATUS             RESTARTS         AGE
csi-snapshotter-0                                                 3/3     Running            0                51m
snapshot-controller-9d578688f-q87jh                               1/1     Running            0                38m
snapshot-controller-9d578688f-tpld5                               1/1     Running            1 (37m ago)      38m
zfs-iscsi-democratic-csi-controller-84655ff96f-7qdtc              6/6     Running            20 (10m ago)     38m
zfs-iscsi-democratic-csi-node-2h5hs                               3/4     CrashLoopBackOff   19 (7s ago)      38m
zfs-iscsi-democratic-csi-node-s46w8                               3/4     CrashLoopBackOff   19 (96s ago)     38m
zfs-nfs-democratic-csi-controller-6cdd6f96fd-m5zz4                6/6     Running            27 (10m ago)     38m
zfs-nfs-democratic-csi-node-hqbb6                                 3/4     CrashLoopBackOff   19 (19s ago)     38m
zfs-nfs-democratic-csi-node-mrsdp                                 3/4     CrashLoopBackOff   19 (9s ago)      38m
zfs-nfs-spitfire-fast-democratic-csi-controller-7496c7fd479995d   6/6     Running            31 (2m40s ago)   38m
zfs-nfs-spitfire-fast-democratic-csi-node-cn8db                   3/4     CrashLoopBackOff   19 (32s ago)     38m
zfs-nfs-spitfire-fast-democratic-csi-node-slgj9                   3/4     CrashLoopBackOff   19 (2s ago)      38m
```

I've identified that the issue seems to be that the kubelet isn't registering my csi drivers but I don't understand what's going wrong. I don't have a non-standard kubelet path and all of the sockets seem to be mounted where they should be.

The driver registrar container of every node seems to just start the registration server and hang as if the kubelet never calls it:

```
❯ k logs -n democratic-csi zfs-nfs-spitfire-fast-democratic-csi-node-cn8db -c driver-registrar
I1126 19:45:40.552849       1 main.go:167] Version: v2.8.0
I1126 19:45:40.552954       1 main.go:168] Running node-driver-registrar in mode=registration
I1126 19:45:40.554292       1 main.go:192] Attempting to open a gRPC connection with: "/csi-data/csi.sock"
I1126 19:45:40.554330       1 connection.go:164] Connecting to unix:///csi-data/csi.sock
I1126 19:45:40.555408       1 main.go:199] Calling CSI driver to discover driver name
I1126 19:45:40.555426       1 connection.go:193] GRPC call: /csi.v1.Identity/GetPluginInfo
I1126 19:45:40.555432       1 connection.go:194] GRPC request: {}
I1126 19:45:40.561196       1 connection.go:200] GRPC response: {"name":"org.democratic-csi.nfs-spitfire-fast","vendor_version":"1.8.4"}
I1126 19:45:40.561217       1 connection.go:201] GRPC error: <nil>
I1126 19:45:40.561228       1 main.go:209] CSI driver name: "org.democratic-csi.nfs-spitfire-fast"
I1126 19:45:40.561266       1 node_register.go:53] Starting Registration Server at: /registration/org.democratic-csi.nfs-spitfire-fast-reg.sock
I1126 19:45:40.561523       1 node_register.go:62] Registration Server started at: /registration/org.democratic-csi.nfs-spitfire-fast-reg.sock
I1126 19:45:40.561606       1 node_register.go:92] Skipping HTTP server because endpoint is set to: ""
```

Kubelet directories and sockets on a node:
```
root@sassaflash:~/ > ls /var/lib/kubelet/plugins_registry/*
/var/lib/kubelet/plugins_registry/org.democratic-csi.iscsi-reg.sock
/var/lib/kubelet/plugins_registry/org.democratic-csi.nfs-reg.sock
/var/lib/kubelet/plugins_registry/org.democratic-csi.nfs-spitfire-fast-reg.sock
root@sassaflash:~/ > ls /var/lib/kubelet/plugins/*         
/var/lib/kubelet/plugins/org.democratic-csi.iscsi:
csi.sock  csi.sock.internal

/var/lib/kubelet/plugins/org.democratic-csi.nfs:
csi.sock  csi.sock.internal

/var/lib/kubelet/plugins/org.democratic-csi.nfs-spitfire-fast:
csi.sock  csi.sock.internal
```

Every pod that depends on pvc provisioned by democratic-csi says:
```
  Warning  FailedMount             54s (x15 over 15m)  kubelet                  MountVolume.MountDevice failed for volume "pvc-2b7b9760-b9b4-43de-9469-075ec190154d" : kubernetes.io/csi: attacher.MountDevice failed to create newCsiDriverClient: driver name org.democratic-csi.nfs not found in the list of registered CSI drivers
  ```
  
  CSI Proxy log:
  ```
  ❯ k logs -n democratic-csi zfs-nfs-democratic-csi-node-ftr9q -c csi-proxy
2023/11/26 19:52:17 BIND_TO [unix:///csi-data/csi.sock], PROXY_TO [unix:///csi-data/csi.sock.internal]
2023/11/26 19:52:17 waiting for socket [/csi-data/csi.sock.internal] to appear, 60s remaining
2023/11/26 19:52:18 waiting for socket [/csi-data/csi.sock.internal] to appear, 59s remaining
2023/11/26 19:52:19 PROXY_TO [unix:///csi-data/csi.sock.internal] is ready!
2023/11/26 19:52:19 BIND_TO [unix:///csi-data/csi.sock] is ready!
2023/11/26 19:52:20 request (://localhost) POST /csi.v1.Identity/GetPluginInfo HTTP/2.0
2023/11/26 19:52:20 request headers map[Content-Type:[application/grpc] Grpc-Timeout:[996440u] Te:[trailers] User-Agent:[grpc-go/1.54.0]]
2023/11/26 19:52:20 Dialing upstream: unix:///csi-data/csi.sock.internal
2023/11/26 19:52:46 request (://localhost) POST /csi.v1.Identity/GetPluginInfo HTTP/2.0
2023/11/26 19:52:46 request headers map[Content-Type:[application/grpc] Grpc-Timeout:[996780u] Te:[trailers] User-Agent:[grpc-go/1.54.0]]
2023/11/26 19:53:16 request (://localhost) POST /csi.v1.Identity/GetPluginInfo HTTP/2.0
2023/11/26 19:53:16 request headers map[Content-Type:[application/grpc] Grpc-Timeout:[996531u] Te:[trailers] User-Agent:[grpc-go/1.54.0]]
2023/11/26 19:53:46 request (://localhost) POST /csi.v1.Identity/GetPluginInfo HTTP/2.0
2023/11/26 19:53:46 request headers map[Content-Type:[application/grpc] Grpc-Timeout:[998111u] Te:[trailers] User-Agent:[grpc-go/1.54.0]]
2023/11/26 19:54:17 request (://localhost) POST /csi.v1.Identity/GetPluginInfo HTTP/2.0
2023/11/26 19:54:17 request headers map[Content-Type:[application/grpc] Grpc-Timeout:[998341u] Te:[trailers] User-Agent:[grpc-go/1.54.0]]
2023/11/26 19:55:38 request (://localhost) POST /csi.v1.Identity/GetPluginInfo HTTP/2.0
2023/11/26 19:55:38 request headers map[Content-Type:[application/grpc] Grpc-Timeout:[996700u] Te:[trailers] User-Agent:[grpc-go/1.54.0]]
```

CSI Driver log:
```
grpc implementation: @grpc/grpc-js
{"host":"soarin","level":"info","message":"initializing csi driver: freenas-nfs","service":"democratic-csi","timestamp":"2023-11-26T19:52:18.127Z"}
{"host":"soarin","level":"info","message":"starting csi server - node version: v16.18.0, package version: 1.8.4, config file: /config/..2023_11_26_19_52_16.43065832/driver-config-file.yaml, csi-name: org.democratic-csi.nfs, csi-driver: freenas-nfs, csi-mode: node, csi-version: 1.5.0, address: , socket: unix:///csi-data/csi.sock.internal","service":"democratic-csi","timestamp":"2023-11-26T19:52:18.759Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: GetPluginInfo call: {\"metadata\":{\"x-forwarded-host\":[\"localhost\"],\"user-agent\":[\"grpc-go/1.54.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:52:20.827Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: GetPluginInfo response: {\"name\":\"org.democratic-csi.nfs\",\"vendor_version\":\"1.8.4\"}","service":"democratic-csi","timestamp":"2023-11-26T19:52:20.827Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: GetPluginInfo call: {\"metadata\":{\"x-forwarded-host\":[\"localhost\"],\"user-agent\":[\"grpc-go/1.54.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:52:46.962Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: GetPluginInfo response: {\"name\":\"org.democratic-csi.nfs\",\"vendor_version\":\"1.8.4\"}","service":"democratic-csi","timestamp":"2023-11-26T19:52:46.962Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: GetPluginInfo call: {\"metadata\":{\"x-forwarded-host\":[\"localhost\"],\"user-agent\":[\"grpc-go/1.54.0\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:53:16.928Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: GetPluginInfo response: {\"name\":\"org.democratic-csi.nfs\",\"vendor_version\":\"1.8.4\"}","service":"democratic-csi","timestamp":"2023-11-26T19:53:16.929Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: Probe call: {\"metadata\":{\"user-agent\":[\"grpc-node-js/1.9.9\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:53:17.049Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: Probe response: {\"ready\":{\"value\":true}}","service":"democratic-csi","timestamp":"2023-11-26T19:53:17.050Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: GetPluginInfo call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.54.0\"],\"x-forwarded-host\":[\"localhost\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:53:46.932Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: GetPluginInfo response: {\"name\":\"org.democratic-csi.nfs\",\"vendor_version\":\"1.8.4\"}","service":"democratic-csi","timestamp":"2023-11-26T19:53:46.933Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: Probe call: {\"metadata\":{\"user-agent\":[\"grpc-node-js/1.9.9\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:54:17.051Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: Probe response: {\"ready\":{\"value\":true}}","service":"democratic-csi","timestamp":"2023-11-26T19:54:17.051Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: GetPluginInfo call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.54.0\"],\"x-forwarded-host\":[\"localhost\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:54:17.983Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: GetPluginInfo response: {\"name\":\"org.democratic-csi.nfs\",\"vendor_version\":\"1.8.4\"}","service":"democratic-csi","timestamp":"2023-11-26T19:54:17.983Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: Probe call: {\"metadata\":{\"user-agent\":[\"grpc-node-js/1.9.9\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:55:17.004Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: Probe response: {\"ready\":{\"value\":true}}","service":"democratic-csi","timestamp":"2023-11-26T19:55:17.004Z"}
{"host":"soarin","level":"info","message":"new request - driver: FreeNASSshDriver method: GetPluginInfo call: {\"metadata\":{\"user-agent\":[\"grpc-go/1.54.0\"],\"x-forwarded-host\":[\"localhost\"]},\"request\":{},\"cancelled\":false}","service":"democratic-csi","timestamp":"2023-11-26T19:55:38.697Z"}
{"host":"soarin","level":"info","message":"new response - driver: FreeNASSshDriver method: GetPluginInfo response: {\"name\":\"org.democratic-csi.nfs\",\"vendor_version\":\"1.8.4\"}","service":"democratic-csi","timestamp":"2023-11-26T19:55:38.697Z"}
```

Kubelet killing driver registrar due to registration not succeeding
```
  Warning  Unhealthy  13m                  kubelet            Liveness probe failed: F1126 19:53:16.663915      38 main.go:160] Kubelet plugin registration hasn't succeeded yet, file=/var/lib/kubelet/plugins/org.democratic-csi.nfs/registration doesn't exist.
  Warning  BackOff    4m1s (x32 over 11m)  kubelet            Back-off restarting failed container driver-registrar in pod zfs-nfs-democratic-csi-node-ftr9q_democratic-csi(13705519-bc08-4397-a68b-0e4ad3d2cec7)
```

Please let me know if I can provide any more useful info to debug this problem!

**Comment by travisghansen** [comment]
Seems like it might be this? https://github.com/democratic-csi/democratic-csi/issues/343

**Comment by starcraft66** [comment]
I don't think so. I took a look at that issue while debugging and I don't see any waiting for sockets. NixOS has a `/etc/localtime` as expected by the helm chart. I just tried setting `node.driver.localtimeHostPath` to `null` as instructed in the linked issue just to make sure and it didn't change anything.

**Comment by travisghansen** [comment]
Hmm, which container is unhealthy? And do you see anything in the logs of the unhealthy container which give an indication of the problem?

**Comment by starcraft66** [comment]
It's the driver-registrar container that crashes constantly, due to its liveness probe failing. Here's a `kubectl describe` of a node agent pod:

```
Name:             zfs-nfs-spitfire-fast-democratic-csi-node-gsn5x
Namespace:        democratic-csi
Priority:         0
Service Account:  zfs-nfs-spitfire-fast-democratic-csi-node-sa
Node:             soarin.235.tdude.co/172.16.29.22
Start Time:       Sun, 26 Nov 2023 21:18:54 -0500
Labels:           app.kubernetes.io/component=node-linux
                  app.kubernetes.io/csi-role=node
                  app.kubernetes.io/instance=zfs-nfs-spitfire-fast
                  app.kubernetes.io/managed-by=Helm
                  app.kubernetes.io/name=democratic-csi
                  controller-revision-hash=7d8cdc5887
                  pod-template-generation=2
Annotations:      checksum/configmap: 541bc7d8a95ba43718f9e1955c282b15fe6370820090d80bcbb2a26c5ed7d47e
                  checksum/secret: 26e07bcb73ff4a7b3083ee9ef5df1c26b08e44338e41f5b082f6476f12219df1
Status:           Running
IP:               172.16.29.22
IPs:
  IP:           172.16.29.22
Controlled By:  DaemonSet/zfs-nfs-spitfire-fast-democratic-csi-node
Containers:
  csi-driver:
    Container ID:  containerd://5dce912b5a71df9075f51efa0c59c6b3402b9bd127e16574bed91417a6f10dcd
    Image:         docker.io/democraticcsi/democratic-csi:latest
    Image ID:      docker.io/democraticcsi/democratic-csi@sha256:1b526bb4e1634ce84534a7a51b038efa91335767c625dcec62cdc42a744e2bcd
    Port:          <none>
    Host Port:     <none>
    Args:
      --csi-version=1.5.0
      --csi-name=org.democratic-csi.nfs-spitfire-fast
      --driver-config-file=/config/driver-config-file.yaml
      --log-level=info
      --csi-mode=node
      --server-socket=/csi-data/csi.sock.internal
    State:          Running
      Started:      Sun, 26 Nov 2023 21:18:56 -0500
    Ready:          True
    Restart Count:  0
    Liveness:       exec [bin/liveness-probe --csi-version=1.5.0 --csi-address=/csi-data/csi.sock.internal] delay=10s timeout=15s period=60s #success=1 #failure=3
    Environment:
      CSI_NODE_ID:           (v1:spec.nodeName)
      NODE_EXTRA_CA_CERTS:  /tmp/certs/extra-ca-certs.crt
    Mounts:
      /config from config (rw)
      /csi-data from socket-dir (rw)
      /dev from dev-dir (rw)
      /host from host-dir (rw)
      /lib/modules from modules-dir (ro)
      /run/udev from udev-data (rw)
      /sys from sys-dir (rw)
      /tmp/certs from extra-ca-certs (rw)
      /var/lib/kubelet from kubelet-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qgtww (ro)
  csi-proxy:
    Container ID:   containerd://3eced253632420b79480d131ba347a257096cffe000ca72a12d74e2b23940ced
    Image:          docker.io/democraticcsi/csi-grpc-proxy:v0.5.3
    Image ID:       docker.io/democraticcsi/csi-grpc-proxy@sha256:4d65ca1cf17d941a8df668b8fe2f1c0cfa512c8b0dbef3ff89a4cd405e076923
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Sun, 26 Nov 2023 21:18:56 -0500
    Ready:          True
    Restart Count:  0
    Environment:
      BIND_TO:   unix:///csi-data/csi.sock
      PROXY_TO:  unix:///csi-data/csi.sock.internal
    Mounts:
      /csi-data from socket-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qgtww (ro)
  driver-registrar:
    Container ID:  containerd://f35b380742a318b1dd1d07abb14239f9a0064c18415608954cb57b6bfa13b31f
    Image:         registry.k8s.io/sig-storage/csi-node-driver-registrar:v2.8.0
    Image ID:      registry.k8s.io/sig-storage/csi-node-driver-registrar@sha256:f6717ce72a2615c7fbc746b4068f788e78579c54c43b8716e5ce650d97af2df1
    Port:          <none>
    Host Port:     <none>
    Args:
      --v=5
      --csi-address=/csi-data/csi.sock
      --kubelet-registration-path=/var/lib/kubelet/plugins/org.democratic-csi.nfs-spitfire-fast/csi.sock
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Tue, 28 Nov 2023 15:11:45 -0500
      Finished:     Tue, 28 Nov 2023 15:12:15 -0500
    Ready:          False
    Restart Count:  836
    Liveness:       exec [/csi-node-driver-registrar --kubelet-registration-path=/var/lib/kubelet/plugins/org.democratic-csi.nfs-spitfire-fast/csi.sock --mode=kubelet-registration-probe] delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      KUBE_NODE_NAME:   (v1:spec.nodeName)
    Mounts:
      /csi-data from socket-dir (rw)
      /registration from registration-dir (rw)
      /var/lib/kubelet from kubelet-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qgtww (ro)
  cleanup:
    Container ID:  containerd://de9390a00d245361d4a51e5fde3ce8516b9c415d5a413f79862ec958bfe27d26
    Image:         docker.io/busybox:1.32.0
    Image ID:      docker.io/library/busybox@sha256:bde48e1751173b709090c2539fdf12d6ba64e88ec7a4301591227ce925f3c678
    Port:          <none>
    Host Port:     <none>
    Command:
      /bin/sh
      -c
      --
    Args:
      while true; do sleep 2; done;
    State:          Running
      Started:      Sun, 26 Nov 2023 21:18:56 -0500
    Ready:          True
    Restart Count:  0
    Environment:    <none>
    Mounts:
      /plugins from plugins-dir (rw)
      /registration from registration-dir (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-qgtww (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True 
Volumes:
  socket-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins/org.democratic-csi.nfs-spitfire-fast
    HostPathType:  DirectoryOrCreate
  plugins-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins
    HostPathType:  Directory
  registration-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins_registry
    HostPathType:  Directory
  kubelet-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet
    HostPathType:  Directory
  dev-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /dev
    HostPathType:  Directory
  modules-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /lib/modules
    HostPathType:  
  udev-data:
    Type:          HostPath (bare host directory volume)
    Path:          /run/udev
    HostPathType:  
  sys-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /sys
    HostPathType:  Directory
  host-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /
    HostPathType:  Directory
  config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  zfs-nfs-spitfire-fast-democratic-csi-driver-config
    Optional:    false
  extra-ca-certs:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      zfs-nfs-spitfire-fast-democratic-csi
    Optional:  false
  kube-api-access-qgtww:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              kubernetes.io/os=linux
Tolerations:                 node.kubernetes.io/disk-pressure:NoSchedule op=Exists
                             node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/network-unavailable:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists
                             node.kubernetes.io/pid-pressure:NoSchedule op=Exists
                             node.kubernetes.io/unreachable:NoExecute op=Exists
                             node.kubernetes.io/unschedulable:NoSchedule op=Exists
Events:
  Type     Reason     Age                      From     Message
  ----     ------     ----                     ----     -------
  Warning  Unhealthy  48m (x2456 over 41h)     kubelet  (combined from similar events): Liveness probe failed: F1128 19:29:05.047665      23 main.go:160] Kubelet plugin registration hasn't succeeded yet, file=/var/lib/kubelet/plugins/org.democratic-csi.nfs-spitfire-fast/registration doesn't exist.
  Normal   Pulled     18m (x832 over 41h)      kubelet  Container image "registry.k8s.io/sig-storage/csi-node-driver-registrar:v2.8.0" already present on machine
  Warning  BackOff    3m16s (x10184 over 41h)  kubelet  Back-off restarting failed container driver-registrar in pod zfs-nfs-spitfire-fast-democratic-csi-node-gsn5x_democratic-csi(04f244f7-d5f1-415d-affc-2246b252ea79)
```

Unfortunately, there is nothing interesting in the driver-registrar container's log, just that the GRPC server for the driver registrar is started.

```
I1128 20:20:55.552040       1 main.go:167] Version: v2.8.0
I1128 20:20:55.552143       1 main.go:168] Running node-driver-registrar in mode=registration
I1128 20:20:55.553530       1 main.go:192] Attempting to open a gRPC connection with: "/csi-data/csi.sock"
I1128 20:20:55.553560       1 connection.go:164] Connecting to unix:///csi-data/csi.sock
I1128 20:20:55.554470       1 main.go:199] Calling CSI driver to discover driver name
I1128 20:20:55.554484       1 connection.go:193] GRPC call: /csi.v1.Identity/GetPluginInfo
I1128 20:20:55.554489       1 connection.go:194] GRPC request: {}
I1128 20:20:55.559925       1 connection.go:200] GRPC response: {"name":"org.democratic-csi.nfs-spitfire-fast","vendor_version":"1.8.4"}
I1128 20:20:55.559938       1 connection.go:201] GRPC error: <nil>
I1128 20:20:55.559947       1 main.go:209] CSI driver name: "org.democratic-csi.nfs-spitfire-fast"
I1128 20:20:55.559979       1 node_register.go:53] Starting Registration Server at: /registration/org.democratic-csi.nfs-spitfire-fast-reg.sock
I1128 20:20:55.560234       1 node_register.go:62] Registration Server started at: /registration/org.democratic-csi.nfs-spitfire-fast-reg.sock
I1128 20:20:55.560319       1 node_register.go:92] Skipping HTTP server because endpoint is set to: ""
```

I did some research on the [Kubelet plugin registration](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins/#device-plugin-registration) mechanism and it seems like it's supposed to call the kubelet and the kubelet would then call this grpc server but nothing's going on presumably.

**Comment by starcraft66** [comment]
Digging through the source code of the kubelet let me better understand the plugin registration mechanism, it turns out that the driver registrar server was never being called because the kubelet discovers plugins with a filesystem watcher and despite what I said earlier, it turned out that I did, in-fact, not have a standard kubelet path. The default being `/var/lib/kubelet` and NixOS's default being `/var/lib/kubernetes`, I got trolled by the `/var/lib/kubelet` path existing due to volume mounts having created it. I ended up cleaning everything up and setting my kubelet path to the standard one in the NixOS module and now my CSI driver is registering. Sorry for the confusion!
